### PR TITLE
Added QA evaluation for GPT & T5

### PR DESCRIPTION
**Pull request by anas-awadalla**
Added a _predict_qa branch in GPT and T5 predict functions.

### CODE REVIEW

File: catwalk/models/gpt.py [hunk 1]
```diff
@@ -1,34 +1,161 @@
...
+            # TODO: Use promptsource to add more prompt options?
+            return instance.context, f"\nQuestion:{instance.question}\nAnswer:"
+
+        converted_instances = Tqdm.tqdm(converted_instances, desc="Processing instances")
+        for batch in more_itertools.chunked(converted_instances, batch_size):
+            formatted_batch = [format_instance(instance) for instance in batch]
+            encodings = tokenizer(formatted_batch,
+                                  padding="max_length",
+                                  truncation="only_first",
+                                  max_length=384,  # TODO: Make this configurable?
```
Comment (jagnusson) [inline catwalk/models/gpt.py:68]: Yah, or perhaps let's just set it from the model config?
```suggestion
                                  max_length=model.config.n_positions - model.config.task_specific_params['text-generation']['max_length'],  # TODO: Make this configurable?
```

File: catwalk/models/gpt.py [hunk 2]
```diff
@@ -1,34 +1,161 @@
...
+
+                    # check if we found a sample that contains the answer and break out of the loop
+                    if len(filtered_encodings["input_ids"])-1 == example_idx:
+                        break
+
+            with torch.inference_mode():
+                outputs = model.generate(input_ids=torch.stack(filtered_encodings["input_ids"]).to(device),
+                                         attention_mask=torch.stack(
+                                             filtered_encodings["attention_mask"]).to(device),
+                                         max_new_tokens=20,
```
Comment (jagnusson) [inline catwalk/models/gpt.py:135]:
```suggestion
                                         max_new_tokens=model.config.task_specific_params['text-generation']['max_length'],
```

File: catwalk/models/gpt.py [hunk 3]
```diff
@@ -1,34 +1,161 @@
...
+
+        def format_instance(instance: HFQAInstance) -> Tuple[str, str]:
+            # TODO: Use promptsource to add more prompt options?
+            return instance.context, f"\nQuestion:{instance.question}\nAnswer:"
+
+        converted_instances = Tqdm.tqdm(converted_instances, desc="Processing instances")
+        for batch in more_itertools.chunked(converted_instances, batch_size):
+            formatted_batch = [format_instance(instance) for instance in batch]
+            encodings = tokenizer(formatted_batch,
+                                  padding="max_length",
```
Comment (jagnusson) [inline catwalk/models/gpt.py:66]: This will allow the model to run a bit faster on shorter inputs now.

```suggestion
                                  padding="longest",
```

File: catwalk/models/gpt.py [hunk 4]
```diff
@@ -1,34 +1,161 @@
...
+                            filtered_encodings["attention_mask"].append(
+                                encodings["attention_mask"][sample_idx])
+                            break # No need to check other answers for this example
+
+                    # check if we found a sample that contains the answer and break out of the loop
+                    if len(filtered_encodings["input_ids"])-1 == example_idx:
+                        break
+
+            with torch.inference_mode():
+                outputs = model.generate(input_ids=torch.stack(filtered_encodings["input_ids"]).to(device),
```
Comment (jagnusson) [inline catwalk/models/gpt.py:132]: This is great! I'll have to add batched generation like this to the EAIGPT model as well.

File: catwalk/models/gpt.py [hunk 5]
```diff
@@ -1,34 +1,161 @@
...
+                                         attention_mask=torch.stack(
+                                             filtered_encodings["attention_mask"]).to(device),
+                                         max_new_tokens=20,
+                                         pad_token_id=tokenizer.pad_token_id,
+                                         num_beams=1)
+            outputs = [output[384 + 1:] for output in outputs]
```
Comment (jagnusson) [inline catwalk/models/gpt.py:138]:
```suggestion
            outputs = [output[len(filtered_encodings["input_ids"][0]) + 1:] for output in outputs]
```

File: catwalk/models/gpt.py [hunk 6]
```diff
@@ -1,34 +1,161 @@
...
+                    # check if we found a sample that contains the answer and break out of the loop
+                    if len(filtered_encodings["input_ids"])-1 == example_idx:
+                        break
+
+            with torch.inference_mode():
+                outputs = model.generate(input_ids=torch.stack(filtered_encodings["input_ids"]).to(device),
+                                         attention_mask=torch.stack(
+                                             filtered_encodings["attention_mask"]).to(device),
+                                         max_new_tokens=20,
+                                         pad_token_id=tokenizer.pad_token_id,
```
Comment (jagnusson) [inline catwalk/models/gpt.py:136]: Only a possibility but we could also have it terminate generation after some token. In the EAI tasks the task itself defines what that token is, but here I guess we'd have to make it a model arg. It's unfortunate those would be different, so I'm not committed to this but I just wanted to paste it here for reference. Feel free to reject the idea.

Also I think integrating this properly would also require purging padding tokens from the generated outputs because of how the batched generation handles the early termination.

```suggestion
                                         pad_token_id=tokenizer.pad_token_id,
                                         eos_token_id=tokenizer.encode('\n')[0],
```

**Review by jagnusson — CHANGES_REQUESTED** [review]
Thanks so much for all this hard work. Your truncation strategy is implemented in such a cool way! I learned a lot from reading this.

I put in a lot of little suggestions, but feel free to follow your best judgement if you disagree. The only major issue is not with the code so much as the possible failure mode of this truncate by windows and search for the answer strategy. That is it seems to fairly often have to drop examples when the answer doesn't fit fully into any of the windows.

I think the simplest resolution for now I'd ask for is if perhaps we could just enable your truncation strategy as an optional behavior and instead default to throwing an error if any input is larger than the max model size. Without ICL support, it shouldn't be common for an input to overflow the max model size anyway I hope.

File: catwalk/models/gpt.py [hunk 7]
```diff
@@ -1,34 +1,164 @@
...
+            encodings = tokenizer(formatted_batch,
+                                  padding="longest",
+                                  truncation="only_first",
+                                  max_length=model.config.n_positions - model.config.task_specific_params['text-generation']['max_length'],  # TODO: Make this configurable?
+                                  stride=128,
+                                  return_overflowing_tokens=True,
+                                  return_offsets_mapping=True,
+                                  return_token_type_ids=True,
+                                  return_tensors="pt"
+                                  )
```
Comment (dirkgr) [inline catwalk/models/gpt.py:74]: With the way windowing works, I might end up with some batches that are much bigger than other batches, right? That doesn't seem ideal. Every instance turns into multiple sequences. Can you make sure that we always run the same number of sequences, and then combine the results for each instance afterwards?

**Review by dirkgr — CHANGES_REQUESTED** [review]

File: catwalk/models/gpt.py [hunk 8]
```diff
@@ -1,34 +1,164 @@
...
+            # Token Type IDs:    0        0       1 (Used to find the last token of the context)
+            # Attention Mask:    0        1       1 (Used to find the first token of the context)
+
+            def get_token_bounds_for_context(sample_idx: int) -> Tuple[int, int]:
+                attention_mask = encodings["attention_mask"][sample_idx]
+                sequence_ids = token_type_map[sample_idx]
+
+                start_index = 0
+                while (attention_mask[start_index] == 0):
+                    start_index += 1
```
[the remark this answers is not in the document]
Comment (anas-awadalla) [reply catwalk/models/gpt.py:99]: Yes I found that doing this improves performance especially if we are not fine-tuning the model like in in-context learning.

File: catwalk/models/gpt.py (hunk 7 above)
Comment (anas-awadalla) [reply catwalk/models/gpt.py:74]: In lines 114 I filter out these encodings into only the first sample for each example that contains an answer. I believe this should ensure that we have the same batch size every time.

Comment (jagnusson) [reply catwalk/models/gpt.py:74]: The batch sizes can still be variable if the answer isn't found in any of the samples for one or more example. I suggested an assert that prevents empty batches, but batches can still vary from full size to less than full size, right?

File: catwalk/models/gpt.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply catwalk/models/gpt.py:99]: Improves quality of the answers, or speed?

File: catwalk/models/gpt.py (hunk 7 above)
Comment (dirkgr) [reply catwalk/models/gpt.py:74]: I am more worried about the opposite case. You want to run a big evaluation (or an evaluation on a small GPU), and so you need to maximize the batch size. But one out of 100 batches ends up 50% bigger than the others, so now you have to make the batch size smaller for the entire evaluation and run at half speed, just to make that one batch fit.

File: catwalk/models/gpt.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (anas-awadalla) [reply catwalk/models/gpt.py:99]: Quality of answer

File: catwalk/models/gpt.py (hunk 7 above)
Comment (anas-awadalla) [reply catwalk/models/gpt.py:74]: Yep this is an issue but we will remove this logic of running multiple samples per example once we switch over to the discussed truncation method.

File: catwalk/models/gpt.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply catwalk/models/gpt.py:99]: I think standard approach is to pad on the right, but adjust position ids? @jagnusson, didn't you do this for OPT?

We want to pretend that each sequence starts at position 0, and that context and continuation are directly adjacent. When the contexts in a batch aren't all the same size, the only way we can do both things is by messing with position ids.

File: catwalk/models/gpt.py (hunk 7 above)
Comment (jagnusson) [reply catwalk/models/gpt.py]: Aha I see, I didn't realize we were discussing the case where we examine all the sampled windows and not just the one with the answer in it. To address @dirkgr's concern, could we try batching by total token count instead of by number of examples in the batch. This would be fussy to implement because you could have to wait for more than one batch to go thru before resolving which of the sampled windows for a given instance wins out. But if it's done right you'll get perfect batch sizes for every batch except the tail one.

File: catwalk/models/gpt.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jagnusson) [reply catwalk/models/gpt.py]: > I think standard approach is to pad on the right, but adjust position ids? @jagnusson, didn't you do this for OPT?
> 
> We want to pretend that each sequence starts at position 0, and that context and continuation are directly adjacent. When the contexts in a batch aren't all the same size, the only way we can do both things is by messing with position ids.

Sorry I didn't see this before! I need to tinker with my notification settings more apparently.

So I unfortunately I don't actually have first hand experience with this. The Metaseq OPT provides its own padding strategy and you just pass it a list of unpadded inputs.

When I looked into this it seemed like [this discussion](https://discuss.huggingface.co/t/batch-generation-with-gpt2/1517) was the definitive one on batched generation in HF. They use padding from the left. I have the same question as @dirkgr though cuz I don't know how they handle the position id's then. I'm not sure if there is an elegant way to pad from the right though because the attention mask may not play nice with trying to mask out padding tokens that fall between the context and the generated continuation (and you still need to adjust the position ids if you want to do that).

File: catwalk/models/gpt.py [hunk 9]
```diff
@@ -1,34 +1,100 @@
...
+                                  padding="longest",
+                                  truncation="only_first",
+                                  max_length=model.config.n_positions - model.config.task_specific_params['text-generation']['max_length'],
+                                  return_tensors="pt")
+            
+            with torch.inference_mode():
+                outputs = model.generate(input_ids=torch.stack(encodings["input_ids"]).to(device),
+                                         attention_mask=torch.stack(
+                                             encodings["attention_mask"]).to(device),
+                                         max_new_tokens=model.config.task_specific_params['text-generation']['max_length'])
```
Comment (dirkgr) [inline catwalk/models/gpt.py:75]: The question remains how padding works here. As @jagnusson points out, regardless of which side you put the padding on, you have to do something with position ids. Or maybe `model.generate()` takes care of that internally? We should know though. In Ian's case, it made a significant difference to performance.

File: catwalk/models/t5.py [hunk 10]
```diff
@@ -21,31 +21,35 @@ def get_model(self) -> T5ForConditionalGeneration:
...
-        task: Task,
-        instances: Sequence[Dict[str, Any]],
-        *,
-        batch_size: int = 32
-    ) -> Iterator[Dict[str, Any]]:
-        if not task.has_instance_conversion(InstanceFormat.T5_PROMPT):
-            raise UnsupportedTaskError(self, task)
+        model = self.get_model().eval()
+        tokenizer = self.get_tokenizer()
+        tokenizer.model_max_length = 1024 # We use this to be similar to the GPT truncation
```
Comment (dirkgr) [inline catwalk/models/t5.py:30]: This has to come from the model or tokenizer though? What if I use a model that has a different max length? Or no max length, like Bloom?

File: catwalk/models/t5.py [hunk 11]
```diff
@@ -21,31 +21,35 @@ def get_model(self) -> T5ForConditionalGeneration:
...
+
+        with torch.inference_mode():
+            for batch in more_itertools.chunked(Tqdm.tqdm(qas, desc="Processing instances"), batch_size):
+                model_input = tokenizer([f"question:{i.question}" for i in batch],
+                                        [f"context:{i.context}" for i in batch],
```
Comment (dirkgr) [inline catwalk/models/t5.py:35]: This has question first. GPT2 has context first. Is this on purpose? Same as a reference implementation?

**Review by dirkgr — CHANGES_REQUESTED** [review]
I just have two questions left. It's possible no code change is needed for them.

File: catwalk/models/t5.py (hunk 10 above)
Comment (anas-awadalla) [reply catwalk/models/t5.py:30]: I believe T5 uses relative attention so it would support longer seq lengths so this is why I went with this fixed length. I think we can just use the 512 in the model config though as I anticipate that a longer default length will be memory consuming.

File: catwalk/models/t5.py (hunk 11 above)
Comment (anas-awadalla) [reply catwalk/models/t5.py:35]: T5 has question first in the HuggingFace implementation. There is no reference for which way to format it when fine-tuning GPT models. I was formatting the input in a similar style as in the GPT3 paper prompts. This makes it trivial to run zero-shot inference (just plug in a pre-trained model) and I believe could simplify some logic when we add in-context learning.

File: catwalk/models/gpt.py (hunk 9 above)
Comment (anas-awadalla) [reply catwalk/models/gpt.py:75]: I looked into this and I believe that this is taken care on inside the generate method using the attention mask as seen [here](https://github.com/huggingface/transformers/blob/21f6f58721dd9154357576be6de54eefef1f1818/src/transformers/models/gpt2/modeling_gpt2.py#L1014). This is also confirmed in this [issue](https://github.com/huggingface/transformers/issues/18478) in a code example from a Huggingface engineer.

File: CHANGELOG.md [hunk 12]
```diff
@@ -21,8 +21,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - SQuAD and SQuADShifts tasks
 - Adds a new MetaICLTask that supports all evaluation tasks in that benchmark
 - Adds a new MetaICLModel that replicates the formatting and truncation used by MetaICL for few shot evaluation
-- Optional random_subsample_seed for PredictStep
-- Adds QA task support for autoregressive models (previously only available with Eleuther task format)
+- Adds QA task support for autoregressive (previously only available with Eleuther task format)
```
Comment (jagnusson) [inline CHANGELOG.md:24]:
```suggestion
- An option for rank classification to average log likelihoods by token length
- Adds QA task support for autoregressive (previously only available with Eleuther task format)
```

**Review by jagnusson — APPROVED** [review]
This looks good to me. There is just one small suggestion I made in the changelog where a line seems to have been dropped. Thanks for all this work!

**Review by dirkgr — APPROVED** [review]
